src/test/java/pingis/cucumber/Stepdefs.java

```diff
@@ -123,37 +198,254 @@ public void not_authenticated() throws Throwable {
+      
+    } else if (verb.equals("join in") || verb.equals("continue")) {
+      beginNewSession();
+      navigateToBaseUrl();
+      is_logged_in(user);
+    }
+  }
 
-  @Given(".*navigates to the task page at (.+)")
-  public void navigates_to_the_task_page_at(String address) throws Throwable {
-    driver.get(baseUrl + address);
+  @Given(".*has chosen to embark on a new challenge")
+  public void navigates_to_the_task_page_at() throws Throwable {
+    findElementByName("Calculator challenge").click();
+    findElementByName("Test task multiply").click();
   }
 
   @Then(".*the page contains (.+)")
   public void page_has_the_right_content(String content) throws Throwable {
     assertTrue(driver.getPageSource().contains(content));
   }
+  
+  @Then(".*the (.+) is shown")
+  public void page_is_shown(String page) throws InterruptedException {
+    String currentUrl = driver.getCurrentUrl();
+    String message = "wrong url, found: " + driver.getCurrentUrl() 
+                   + ",  \n expected to contain " + testUrls.get(page);
+    
+    assertTrue(message, driver.getCurrentUrl().contains(testUrls.get(page)));
+    assertTrue(isDisplayed(page));
+  }
+  
+  @And("(.+) has successfully submitted new challenge with first task pair")
+  public void has_successfully_submitted_new_challenge_with_first_task_pair(String user) 
+                                                throws InterruptedException, Throwable {
+    wants_to_add_verb_here_live_challenge(user, "participate in");
+    inputs_and_submits_live_challenge_description();
+    inputs_and_submits_new_task_pair("first");
+  }
+  
+  @And("(.+) has successfully submitted (.+) code for (.+) task pair")
+  public void has_successfully_submitted_test_code_for_first_task(String user, String codetype, 
+                                                                  String howMany) 
+                                                throws InterruptedException {
+    inputs_and_submits_code_to_editor(user, codetype, howMany);
+    waitForFeedbackSuccessMessage();
+  }
+
+  private void waitForFeedbackSuccessMessage() throws InterruptedException {
+    WebElement myDynamicElement = (new WebDriverWait(driver, 12))
+            .until(ExpectedConditions.presenceOfElementLocated(By.id("results-panel")));
+  }
+
+  @And("page (.+) the (.+) input by (.+)")
+  public void page_contains_the_code_input_by(String contains, String codeType, String user) {
+    boolean isContains = contains.equals("contains"); 
+    assertEquals(isContains, isDisplayed(codeType + " input by " + user));
+  }
+  
+  @When("(.+) inputs and submits (.+) code to (.+) task pair")
+  public void inputs_and_submits_code_to_editor(String user, String codetype, String howMany) 
+                                                throws InterruptedException {
+    if (!codetype.equals("empty")) { 
+      input_to_editor(inputs.get(howMany + " " + user + " " + codetype + " task code"));
```

Comment:
Confusing. How about using the name of the first tab as a key? Selenium elements should have a `getText()` or equivalent.